core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java

```diff
@@ -252,22 +260,44 @@ private void deleteFromStoresAndNotify(K key, V value, Metadata metadata) {
             cacheNotifier.notifyCacheEntryExpired(key, value, metadata, null)));
    }
 
-   protected Long localLastAccess(Object key, Object value, int segment) {
-      InternalCacheEntry ice = dataContainer.running().peek(segment, key);
-      if (ice != null && (value == null || value.equals(ice.getValue())) &&
-            !ice.isExpired(timeService.wallClockTime())) {
-         return ice.getLastUsed();
+   @Override
+   public CompletionStage<Boolean> handlePossibleExpiration(InternalCacheEntry<K, V> ice, int segment, boolean isWrite) {
+      long currentTime = timeService.wallClockTime();
+      if (ice.isExpired(currentTime)) {
+         if (trace) {
+            log.tracef("Retrieved entry for key %s was expired locally, attempting expiration removal", ice.getKey());
+         }
+         CompletableFuture<Boolean> expiredStage = entryExpiredInMemory(ice, currentTime, isWrite);
+         if (trace) {
+            expiredStage = expiredStage.thenApply(expired -> {
+               if (expired == Boolean.FALSE) {
+                  log.tracef("Retrieved entry for key %s was found to not be expired.", ice.getKey());
+               } else {
+                  log.tracef("Retrieved entry for key %s was confirmed to be expired.", ice.getKey());
+               }
+               return expired;
+            });
+         }
+         return expiredStage;
+      } else if (!isWrite && ice.canExpireMaxIdle()) {
+         return touchEntryAndReturnIfExpired(ice, segment);
       }
-      return null;
+      return CompletableFutures.completedFalse();
    }
 
-   @Override
-   public CompletableFuture<Long> retrieveLastAccess(Object key, Object value, int segment) {
-      Long lastAccess = localLastAccess(key, value, segment);
-      if (lastAccess != null) {
-         return CompletableFuture.completedFuture(lastAccess);
-      }
-      return CompletableFutures.completedNull();
+   /**
+    * Response is whether the value should be treated as expired - thus if both local and remote were able to touch
```

Comment:
IMO we shouldn't mention remote invocations here, just the first part, since `ExpirationManagerImpl` doesn't go remotely.